### PR TITLE
Raise an error if torch.cat is given `out` as one of the input tensors

### LINKED ISSUE #30562: `torch.cat` probably modified after being freed
## 🐛 Use After Free Bug in `torch.cat`

When using `torch.cat` with zero sized arrays and an out parameter, `torch.cat` produces a memory error:

```
Python(2479,0x11b476dc0) malloc: Incorrect checksum for freed object 0x7f9834954060: probably modified after being freed.
Corrupt value: 0x3f61f243bf3752ad
```

## To Reproduce

1. Run this code:

```python
import torch
z = torch.zeros((0))
r = torch.randn(4, 6)
torch.cat([z, r], dim=0, out=z)
```

## Expected behavior

While this may be an incorrect use of `cat` (I'm not really sure if it is incorrect), it should still perform necessary dynamic checks so as not to expose a memory error.

## Environment

Collecting environment information...
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.15.1
GCC version: Could not collect
CMake version: version 3.14.3

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.15.3
[pip3] torch==1.1.0
[pip3] torchvision==0.3.0
[pip3] torchviz==0.0.1
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @jerryzh168

### PR DESCRIPTION
**Pull request by peterbell10**
Fixes #30562 for both cpu and cuda.

### CODE REVIEW

**Comment by ezyang** [comment]
Testing tensors for object equality is generally a smell for me, because it is almost never what you actually want. Can you explain the root cause of the error in the issue, and why you think this is the most appropriate fix?

**Comment by peterbell10** [comment]
As I understand it, the issue is that the `result` tensor is resized to the full output size; then its full size is copied into itself, in addition to the other tensors, causing a buffer overrun. I think this only happens if `result` is the same tensor object since the input tensor needs to have its shape updated when the output tensor is resized.

**Comment by ezyang** [comment]
OK. The diagnosis you present seems plausible. However, your test is insufficient. Suppose that I run:

```
x = torch.zeros(2)
y = torch.ones(2)
print(torch.cat(y, x, out=x.view(2)))
```

`x.view(2)` returns a distinct Tensor that is a view of x, so that the identity of x and x.view(2) are not the same (equality test will fail.) But y will still get written into the memory of x, clobbering the original contents of x, so you'll end up with two y's cat'ed.

I get

```
tensor([1., 1., 1., 1.])
```

which is wrong.

**Comment by peterbell10** [comment]
That's fair. I was thinking mainly of fixing the crash but there are certainly correctness implications as well. I'll have another think about this next week.

**Comment by peterbell10** [comment]
Alright, so instead of tensor object equality we could check for storage object equality. This would detect the case you suggested but would also error on some working calls such as:
```python
>>> x = torch.arange(10)
>>> torch.cat(x[:2], x[4:6], out=x[6:])
tensor([0, 1, 4, 5])
>>> x
tensor([0, 1, 2, 3, 4, 5, 0, 1, 4, 5])
```

So, we could either:
- Only error when the memory directly overlaps. Easy to implement for contiguous tensors but harder for strided views. Only errors on code that was broken already. 
- Allocate a new storage object for the output when it may overlap with input tensors. Never errors so no user code is broken but it may use more memory.
- Always error on storage object equality. A breaking change, but probably a rare usage scenario.

**Comment by albanD** [comment]
I guess the full solution would be that in the case of same storage, we run the expensive overlap detection code. And raise an error when we detect overlap between an input and the output.
The should allow all use cases without any regression.
But I don't know if we currently have a nice function that can do overlap detection?

**Comment by peterbell10** [comment]
There is a `get_overlap_status` function:
https://github.com/pytorch/pytorch/blob/174e1ba3b8ed640c09b5ff927227e765193d0cad/aten/src/ATen/MemoryOverlap.cpp#L41

However, it just gives up if the tensors are strided:
https://github.com/pytorch/pytorch/blob/174e1ba3b8ed640c09b5ff927227e765193d0cad/aten/src/ATen/MemoryOverlap.cpp#L43-L45

**Comment by albanD** [comment]
I would personally lean towards `Always error on storage object equality. A breaking change, but probably a rare usage scenario.` then. But are we sure that only a single Storage can look at a given memory at once? (with distributed, shared memory, etc) If this is not the case, then the `Storage` equality won't be a good check either.

**Comment by peterbell10** [comment]
I had a look at what `TensorIterator` does and it only errors if `get_overlap_status(a, b) == MemOverlapStatus::PARTIAL`. So it doesn't attempt to protect against strided tensor overlaps.

**Comment by albanD** [comment]
I think we would be happy here to provide the same guarantees as the Tensor Iterator ones.
@gchanan does that sounds good to you?

**Comment by ezyang** [comment]
Last time we talked about this: https://github.com/pytorch/pytorch/issues/8212

I think following TensorIterator is good, given that's where @pbelevich put his fix

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit e92b970:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=e92b970158fcce602dd3d9ffb687981afa72b6ac).

### 1 new failure recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/3928522) pytorch_linux_xenial_py3_6_gcc5_4_build (1/1)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=73446608))
```
CONFLICT (add/add): Merge conflict in .circleci/scripts/binary_checkout.sh
Auto-merging .circleci/scripts/binary_checkout.sh
CONFLICT (add/add): Merge conflict in .circleci/generate_config_yml.py
Auto-merging .circleci/generate_config_yml.py
CONFLICT (add/add): Merge conflict in .circleci/config.yml
Auto-merging .circleci/config.yml
CONFLICT (add/add): Merge conflict in .circleci/cimodel/data/caffe2_build_definitions.py
Auto-merging .circleci/cimodel/data/caffe2_build_definitions.py
CONFLICT (add/add): Merge conflict in .circleci/cimodel/data/binary_build_definitions.py
Auto-merging .circleci/cimodel/data/binary_build_definitions.py
Automatic merge failed; fix conflicts and then commit the result.
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Comment by gchanan** [comment]
sgtm.

**Comment by peterbell10** [comment]
Looks like the build is now failing after rebasing on the latest master. If I understand correctly, the splitting of `torch_cpu` and `torch_cuda` in #30315 means I can't use `at::get_overlap_status` in `THC` code. I'm not sure what my options are here; is there any way to keep using it?

It also looks like the stricter memory checks are causing test failures. Even more curious, these test aren't even using the `out` parameter:
```
RuntimeError: invalid argument 0: unsupported operation: the input tensors cannot refer to any of the output memory locations. Found overlap in input tensor 0. at /var/lib/jenkins/workspace/aten/src/TH/generic/THTensor.cpp:634
The above operation failed in interpreter.
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/jit/test_models.py", line 467
                    h_t2, c_t2 = self.lstm2(h_t, (h_t2, c_t2))
                    output = self.linear(h_t2)
                    outputs = torch.cat((outputs, output), 1)
                              ~~~~~~~~~ <--- HERE
```

I'm not familiar with the JIT but could it be rewriting `x = f(y)` as `f(y, out=x)`?

**Comment by ezyang** [comment]
> I can't use at::get_overlap_status in THC code. I'm not sure what my options are here; is there any way to keep using it?

Yes, mark it `TORCH_API` and it will be visible.

**Comment by ezyang** [comment]
As for the error in question, are you handling zero size tensors correctly?

**Comment by peterbell10** [comment]
Thanks for the insight @ezyang, now the only remaining test failure is `test_cat_out_xla`. Should I just skip that test on XLA?

**Comment by ezyang** [comment]
I guess? cc @ailzhang it's possible XLA has the same problem, but maybe not.

**Comment by ailzhang** [comment]
Yea feel free to send a PR in `pytorch/xla` to skip the test for now(a simple example is https://github.com/pytorch/xla/pull/1472). 
We can fix the problem later. cc: @mruberry for visibility :D

**Comment by peterbell10** [comment]
The one test failure is just a flake in `TestMomentumSGD.test_fp16momentum_sgd`:
```
21:58:20     def __flaky(self, message):
21:58:20         if len(self.falsifying_examples) <= 1:
21:58:20 >           raise Flaky(message)
21:58:20 E           hypothesis.errors.Flaky: Hypothesis test_fp16momentum_sgd(self=<caffe2.python.operator_test.momentum_sgd_test.TestMomentumSGD testMethod=test_fp16momentum_sgd>, n=7, nesterov=True, gc=device_type: 6, dc=[, device_type: 6]) produces unreliable results: Falsified on the first call but did not on a subsequent one
```

File: aten/src/TH/generic/THTensor.cpp [hunk 1]
```diff
@@ -623,6 +624,16 @@ void THTensor_(catArray)(THTensor *result, THTensor **inputs, int numInputs, int
   bool allSkipped= true;
   int64_t nDims = 0;
   THTensor *notSkippedTensor;  // non-owning reference
+
+  // Inputs cannot alias the output tensor
+  for (int i = 0; i < numInputs; i++) {
+    auto lap = at::get_overlap_status(result, inputs[i]);
+    THArgCheck(lap != at::MemOverlapStatus::PARTIAL &&
```
Comment (ezyang) [inline aten/src/TH/generic/THTensor.cpp]: nit: use the new hotness `TORCH_CHECK`

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**